### PR TITLE
Make git operations support committish

### PR DESCRIPTION
**Pull request by yufeih**
This PR makes all git operations support committish: branch, tags, short hash, etc... Unblocks deployment due to impact test uses commit hash in dependency configuration.

- Make GetCommitHistory, Fetch support committish (fetch will first attempt to fetch single branch, then fallback to fetch all tags and branches)
- Split stdout and stderr from process exec, to detect fetch errors while preserving fetch progress
- Adjust a test case to use short hash

### CODE REVIEW

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 1]
```diff
@@ -228,45 +228,43 @@ public void Dispose()
...
-            }
+            git_revwalk_push(walk, git_object_id(headCommit));
+            git_object_free(headCommit);
 
             while (true)
             {
                 var error = git_revwalk_next(out var commitId, walk);
                 if (error == -31 /* GIT_ITEROVER */)
+                {
                     break;
```
Comment (live1206) [inline src/docfx/lib/git/GitCommitProvider.cs]: should `walk` be freed while exiting as well?

File: src/docfx/lib/git/GitUtility.cs [hunk 2]
```diff
@@ -174,10 +174,30 @@ public static Task PruneWorkTrees(string cwd)
...
+            string result = null;
+
+            if (string.IsNullOrEmpty(committish))
+            {
+                committish = "HEAD";
+            }
+
+            if (git_repository_open(out var repo, repoPath) != 0)
+            {
+                throw new InvalidOperationException($"Not a git repo {repoPath}");
```
Comment (live1206) [inline src/docfx/lib/git/GitUtility.cs]: should `repo` be freed before throwing?

File: src/docfx/lib/git/GitUtility.cs [hunk 3]
```diff
@@ -174,10 +174,30 @@ public static Task PruneWorkTrees(string cwd)
...
+            }
+
+            if (git_revparse_single(out var reference, repo, committish) == 0)
+            {
+                result = git_object_id(reference)->ToString();
+                git_object_free(reference);
```
Comment (live1206) [inline src/docfx/lib/git/GitUtility.cs]: free `reference` outside `if` block?

File: src/docfx/lib/git/GitCommitProvider.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/lib/git/GitCommitProvider.cs]: it is freed after while loop

File: src/docfx/lib/git/GitUtility.cs (hunk 3 above)
Comment (yufeih) [reply src/docfx/lib/git/GitUtility.cs]: when `git_revparse_single` fails (returns non zero), reference is an invalid point, you can only free a valid pointer.

File: src/docfx/lib/git/GitUtility.cs (hunk 2 above)
Comment (yufeih) [reply src/docfx/lib/git/GitUtility.cs]: nope

File: src/docfx/lib/git/GitCommitProvider.cs [hunk 4]
```diff
@@ -231,45 +228,43 @@ public void Dispose()
             }
         }
 
-        private unsafe (List<Commit>, Dictionary<long, Commit>) LoadCommits(string branchName = null)
+        private unsafe (List<Commit>, Dictionary<long, Commit>) LoadCommits(string committish = null)
         {
```
Comment (OsmondJiang) [inline src/docfx/lib/git/GitCommitProvider.cs]: >committish [](start = 63, length = 10)

how about the local and remote?

File: test/docfx.Test/lib/GitUtilityTest.cs [hunk 5]
```diff
@@ -59,16 +59,6 @@ public static async Task GetCommitsSameAsGitExe(string file)
             }
         }
 
-        [Fact]
-        public static async Task GitCommandConcurreny()
-        {
-            var cwd = GitUtility.FindRepo(Path.GetFullPath("README.md"));
```
Comment (OsmondJiang) [inline test/docfx.Test/lib/GitUtilityTest.cs]: >var cwd = GitUtility.FindRepo(Path.GetFullPath("README.md")); [](start = 12, length = 61)

why remove this test?

Comment (yufeih) [reply test/docfx.Test/lib/GitUtilityTest.cs]: moved to ProcessUtilityTest

File: test/docfx.Test/lib/GitUtilityTest.cs [hunk 6]
```diff
@@ -48,8 +48,8 @@ public static async Task GetCommitsSameAsGitExe(string file)
                     string.Join("\n", lib.Select(c => $"{c.Sha}|{c.Time.ToString("s")}{c.Time.ToString("zzz")}|{c.AuthorName}|{c.AuthorEmail}")));
 
                 // another branch
-                exe = Exec("git", $"--no-pager log --format=\"%H|%cI|%an|%ae\" origin/test -- \"{pathToRepo}\"", repo);
-                lib = commitsProvider.GetCommitHistory(pathToRepo, "test");
+                exe = Exec("git", $"--no-pager log --format=\"%H|%cI|%an|%ae\" a050eaf -- \"{pathToRepo}\"", repo);
```
Comment (OsmondJiang) [inline test/docfx.Test/lib/GitUtilityTest.cs]: >a050eaf [](start = 79, length = 7)

also add a normal other branch name test?

Comment (yufeih) [reply test/docfx.Test/lib/GitUtilityTest.cs]: this test fails if your local branch does not have the same commit as server, or if origin is a forked repo that does not have a test branch (like me), a commit hash will always have a the same result.

**Pull request merged**